## Datashader doesn't support PySpark dataframe, so moved to Dask that is supported and can be deployed in the cluster

In [1]:
from dask.distributed import Client, progress
client = Client('127.0.0.1:8786')
client

<Client: scheduler='tcp://127.0.0.1:8786' processes=4 cores=8>

### Connect to Hive table via pyspark.sql HiveContext and fetch only dropoff/pickup lon/lat columns

In [29]:
import findspark
findspark.init()

import pyspark
from pyspark import SparkContext
from pyspark.sql import HiveContext

sc = SparkContext(appName="taxi with Dask")
hiveContxt = HiveContext(sc)

# sc.stop()

In [30]:
query = "SELECT pickup_longitude, pickup_latitude, dropoff_longitude, dropoff_latitude FROM green_taxi"
results = hiveContxt.sql(query)

In [31]:
results.cache()

DataFrame[pickup_longitude: decimal(9,6), pickup_latitude: decimal(9,6), dropoff_longitude: decimal(9,6), dropoff_latitude: decimal(9,6)]

In [32]:
results.count()

52667191

In [33]:
results = results.filter((results.pickup_longitude != 0) & \
                         (results.pickup_latitude != 0) & \
                         (results.dropoff_longitude != 0) & \
                         (results.dropoff_latitude != 0))
results.count()

52447825

#### There are 52667191- 52447825 = 219366 zeros rows

### Function to convert lon/lat to web mercator format to plot

In [34]:
from pyspark.sql.types import FloatType

# Change dtype of the DataFrame to avoid operand type error with the UDF
results = results.withColumn("pickup_longitude", results["pickup_longitude"].cast(FloatType()))
results = results.withColumn("pickup_latitude", results["pickup_latitude"].cast(FloatType()))
results = results.withColumn("dropoff_longitude", results["dropoff_longitude"].cast(FloatType()))
results = results.withColumn("dropoff_latitude", results["dropoff_latitude"].cast(FloatType()))

results.dtypes

[('pickup_longitude', 'float'),
 ('pickup_latitude', 'float'),
 ('dropoff_longitude', 'float'),
 ('dropoff_latitude', 'float')]

In [35]:
from pyspark.sql.functions import udf
import math


def toWebMercatorLon(xLon):
    # Check if coordinate out of range for Latitude/Longitude
    if abs(xLon) > 180:
        return
 
    semimajorAxis = 6378137.0  # WGS84 spheriod semimajor axis
    east = xLon * 0.017453292519943295
 
    easting = semimajorAxis * east
 
    return easting


def toWebMercatorLat(yLat):
    # Check if coordinate out of range for Latitude/Longitude
    if abs(yLat) > 90:
        return
 
    semimajorAxis = 6378137.0  # WGS84 spheriod semimajor axis
    north = yLat * 0.017453292519943295
 
    northing = 3189068.5 * math.log((1.0 + math.sin(north)) / (1.0 - math.sin(north)))
 
    return northing

In [36]:
toWebMercatorLon_udf = udf(toWebMercatorLon, FloatType())  # FloatType is the the type returned by the function
toWebMercatorLat_udf = udf(toWebMercatorLat, FloatType())

results = results.withColumn("pickup_longitude", toWebMercatorLon_udf("pickup_longitude")) 
results = results.withColumn("pickup_latitude", toWebMercatorLat_udf("pickup_latitude")) 
results = results.withColumn("dropoff_longitude", toWebMercatorLon_udf("dropoff_longitude")) 
results = results.withColumn("dropoff_latitude", toWebMercatorLat_udf("dropoff_latitude")) 

In [37]:
results.show(10)

+----------------+---------------+-----------------+----------------+
|pickup_longitude|pickup_latitude|dropoff_longitude|dropoff_latitude|
+----------------+---------------+-----------------+----------------+
|      -8230714.5|      4976782.5|       -8230714.5|       4976782.5|
|      -8230732.5|      4976689.0|       -8230732.5|       4976689.0|
|    -1.2821679E7|      4319502.0|     -1.2821705E7|       4319445.5|
|    -1.2821679E7|      4319502.0|     -1.2821705E7|       4319445.5|
|     -1.282169E7|      4319495.0|      -1.282169E7|       4319490.0|
|    -1.2821705E7|      4319454.5|     -1.2821698E7|       4319464.5|
|      -8230709.5|      4976771.5|       -8230709.5|       4976766.5|
|      -8569337.0|      4711232.0|       -8569343.0|       4711267.0|
|      -8230706.0|      4976773.5|       -8230717.0|       4976764.5|
|      -8230719.0|      4976770.0|       -8230715.5|       4976778.0|
+----------------+---------------+-----------------+----------------+
only showing top 10 

### Export the dataframe to parquet file to further process with Dask dataframe

In [38]:
%%time
results.write.parquet("s3a://nyctaxidataset/parquet_taxi/green_taxi_coord.parquet")

Py4JJavaError: An error occurred while calling o157.parquet.
: org.apache.spark.SparkException: Job aborted.
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation$$anonfun$run$1.apply$mcV$sp(InsertIntoHadoopFsRelation.scala:154)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation$$anonfun$run$1.apply(InsertIntoHadoopFsRelation.scala:106)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation$$anonfun$run$1.apply(InsertIntoHadoopFsRelation.scala:106)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:53)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation.run(InsertIntoHadoopFsRelation.scala:106)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult$lzycompute(commands.scala:58)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult(commands.scala:56)
	at org.apache.spark.sql.execution.ExecutedCommand.doExecute(commands.scala:70)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:132)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:130)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:130)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:56)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:56)
	at org.apache.spark.sql.execution.datasources.ResolvedDataSource$.apply(ResolvedDataSource.scala:256)
	at org.apache.spark.sql.DataFrameWriter.dataSource$lzycompute$1(DataFrameWriter.scala:181)
	at org.apache.spark.sql.DataFrameWriter.org$apache$spark$sql$DataFrameWriter$$dataSource$1(DataFrameWriter.scala:181)
	at org.apache.spark.sql.DataFrameWriter$$anonfun$save$1.apply$mcV$sp(DataFrameWriter.scala:188)
	at org.apache.spark.sql.DataFrameWriter.executeAndCallQEListener(DataFrameWriter.scala:154)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:188)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:172)
	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:370)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)
	at py4j.Gateway.invoke(Gateway.java:259)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:209)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 238 in stage 5.0 failed 4 times, most recent failure: Lost task 238.3 in stage 5.0 (TID 766, ip-172-31-39-202.ec2.internal, executor 14): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 111, in main
    process()
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 106, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 263, in dump_stream
    vs = list(itertools.islice(iterator, batch))
  File "/usr/lib/spark/python/pyspark/sql/functions.py", line 1566, in <lambda>
    func = lambda _, it: map(lambda x: returnType.toInternal(f(*x)), it)
  File "<ipython-input-35-b0a78c3f6140>", line 26, in toWebMercatorLat
ZeroDivisionError: float division by zero

	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRDD.scala:166)
	at org.apache.spark.api.python.PythonRunner$$anon$1.next(PythonRDD.scala:129)
	at org.apache.spark.api.python.PythonRunner$$anon$1.next(PythonRDD.scala:125)
	at org.apache.spark.InterruptibleIterator.next(InterruptibleIterator.scala:43)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$GroupedIterator.takeDestructively(Iterator.scala:913)
	at scala.collection.Iterator$GroupedIterator.go(Iterator.scala:929)
	at scala.collection.Iterator$GroupedIterator.fill(Iterator.scala:969)
	at scala.collection.Iterator$GroupedIterator.hasNext(Iterator.scala:972)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at org.apache.spark.api.python.PythonRDD$.writeIteratorToStream(PythonRDD.scala:452)
	at org.apache.spark.api.python.PythonRunner$WriterThread$$anonfun$run$3.apply(PythonRDD.scala:280)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1819)
	at org.apache.spark.api.python.PythonRunner$WriterThread.run(PythonRDD.scala:239)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1457)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1445)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1444)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1444)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
	at scala.Option.foreach(Option.scala:236)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:799)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1668)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1627)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1616)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:620)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1862)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1875)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1952)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation$$anonfun$run$1.apply$mcV$sp(InsertIntoHadoopFsRelation.scala:148)
	... 32 more
Caused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 111, in main
    process()
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 106, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 263, in dump_stream
    vs = list(itertools.islice(iterator, batch))
  File "/usr/lib/spark/python/pyspark/sql/functions.py", line 1566, in <lambda>
    func = lambda _, it: map(lambda x: returnType.toInternal(f(*x)), it)
  File "<ipython-input-35-b0a78c3f6140>", line 26, in toWebMercatorLat
ZeroDivisionError: float division by zero

	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRDD.scala:166)
	at org.apache.spark.api.python.PythonRunner$$anon$1.next(PythonRDD.scala:129)
	at org.apache.spark.api.python.PythonRunner$$anon$1.next(PythonRDD.scala:125)
	at org.apache.spark.InterruptibleIterator.next(InterruptibleIterator.scala:43)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$GroupedIterator.takeDestructively(Iterator.scala:913)
	at scala.collection.Iterator$GroupedIterator.go(Iterator.scala:929)
	at scala.collection.Iterator$GroupedIterator.fill(Iterator.scala:969)
	at scala.collection.Iterator$GroupedIterator.hasNext(Iterator.scala:972)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at org.apache.spark.api.python.PythonRDD$.writeIteratorToStream(PythonRDD.scala:452)
	at org.apache.spark.api.python.PythonRunner$WriterThread$$anonfun$run$3.apply(PythonRDD.scala:280)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1819)
	at org.apache.spark.api.python.PythonRunner$WriterThread.run(PythonRDD.scala:239)


### Load the parquet file into a Dask dataframe

In [2]:
import dask.dataframe as dd
from dask.distributed import progress
df = dd.read_parquet('s3://nyctaxidataset/parquet_taxi/green_taxi.parquet',
                storage_options={'key':'AKIAIHCSMTKCVZ33ZGJA', 
                                 'secret':'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'})


df = client.persist(df)
progress(df)


In [3]:
df.head(150)

pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude
0                 NaN              NaN                NaN               NaN
1                 NaN              NaN                NaN               NaN
2            0.000000         0.000000           0.000000          0.000000
3            0.000000         0.000000           0.000000          0.000000
4            0.000000         0.000000           0.000000          0.000000
5            0.000000         0.000000           0.000000          0.000000
6            0.000000         0.000000           0.000000          0.000000
7            0.000000         0.000000           0.000000          0.000000
8            0.000000         0.000000           0.000000          0.000000
9          -73.937767        40.758480         -73.937767         40.758480
10         -73.937927        40.757843         -73.937927         40.757843
11           0.000000         0.000000           0.000000          0.000000
12           0.000000         0.000000           0.000000          0.000000
13           0.000000         0.000000           0.000000          0.000000
14           0.000000         0.000000           0.000000          0.000000
15           0.000000         0.000000           0.000000          0.000000
16           0.000000         0.000000           0.000000          0.000000
17           0.000000         0.000000           0.000000          0.000000
18           0.000000         0.000000           0.000000          0.000000
19           0.000000         0.000000           0.000000          0.000000
20           0.000000         0.000000           0.000000          0.000000
21           0.000000         0.000000           0.000000          0.000000
22           0.000000         0.000000           0.000000          0.000000
23           0.000000         0.000000           0.000000          0.000000
24           0.000000         0.000000           0.000000          0.000000
25           0.000000         0.000000           0.000000          0.000000
26           0.000000         0.000000           0.000000          0.000000
27           0.000000         0.000000           0.000000          0.000000
28           0.000000         0.000000           0.000000          0.000000
29           0.000000         0.000000           0.000000          0.000000
..                ...              ...                ...               ...
120          0.000000         0.000000           0.000000          0.000000
121          0.000000         0.000000           0.000000          0.000000
122          0.000000         0.000000           0.000000          0.000000
123          0.000000         0.000000           0.000000          0.000000
124          0.000000         0.000000           0.000000          0.000000
125          0.000000         0.000000           0.000000          0.000000
126          0.000000         0.000000           0.000000          0.000000
127          0.000000         0.000000           0.000000          0.000000
128          0.000000         0.000000           0.000000          0.000000
129          0.000000         0.000000           0.000000          0.000000
130          0.000000         0.000000           0.000000          0.000000
131          0.000000         0.000000           0.000000          0.000000
132          0.000000         0.000000           0.000000          0.000000
133          0.000000         0.000000           0.000000          0.000000
134          0.000000         0.000000           0.000000          0.000000
135        -73.937721        40.758404         -73.937721         40.758369
136          0.000000         0.000000           0.000000          0.000000
137        -76.979668        38.927082         -76.979713         38.927326
138          0.000000         0.000000           0.000000          0.000000
139          0.000000         0.000000           0.000000          0.000000
140        -73.937691        40.758419         -73.93

In [4]:
df_mercator = df[(df.pickup_longitude != 0) & \
        (df.pickup_latitude != 0) & \
        (df.dropoff_longitude != 0) & \
        (df.dropoff_latitude != 0)]
df_mercator = df_mercator.compute()
df_mercator.head(150)

distributed.client - WARNING - Client report stream closed to scheduler
distributed.utils - ERROR - in <closed TCP>: Stream is closed: while trying to call remote method 'gather'
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.5/site-packages/distributed/comm/tcp.py", line 170, in read
    n_frames = yield stream.read_bytes(8)
  File "/home/ubuntu/anaconda3/lib/python3.5/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/ubuntu/anaconda3/lib/python3.5/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
tornado.iostream.StreamClosedError: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.5/site-packages/distributed/core.py", line 454, in send_recv_from_rpc
    result = yield send_recv(comm=comm, op=key, **kwargs)
  File 

CommClosedError: in <closed TCP>: Stream is closed: while trying to call remote method 'gather'